orc8r/cloud/go/syncstore/cache_writer.go

```
@@ -0,0 +1,145 @@
+/*
+ Copyright 2020 The Magma Authors.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree.
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package syncstore
+
+import (
+	"database/sql"
+	"fmt"
+
+	"magma/orc8r/cloud/go/sqorc"
+
+	"github.com/Masterminds/squirrel"
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+)
+
+type cacheWriter struct {
+	network       string
+	id            string
+	db            *sql.DB
+	builder       sqorc.StatementBuilder
+	invalidWriter bool
+}
+
+func NewCacheWriter(network string, id string, db *sql.DB, builder sqorc.StatementBuilder) CacheWriter {
+	return &cacheWriter{network: network, id: id, db: db, builder: builder, invalidWriter: false}
+}
+
+// InsertMany inserts a batch of objects into the temporary table of the
+// CacheWriter object.
+func (l *cacheWriter) InsertMany(objects map[string][]byte) error {
+	if l.invalidWriter {
+		return errors.Errorf("attempt to insert into network %+v with invalid cache writer", l.network)
+	}
+	if len(objects) == 0 {
+		return nil
+	}
+
+	insertQuery := l.builder.
+		Insert(l.id).
+		Columns(nidCol, idCol, objCol)
+	errs := &multierror.Error{}
+	for id, obj := range objects {
+		insertQuery = insertQuery.Values(l.network, id, obj)
+	}
+	if errs.ErrorOrNil() != nil {
+		return errs
+	}
+
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		_, err := insertQuery.RunWith(tx).Exec()
+		return nil, errors.Wrapf(err, "insert objs into store for network %+v", l.network)
+	}
+	_, err := sqorc.ExecInTx(l.db, nil, nil, txFn)
+	return err
+}
+
+func (l *cacheWriter) Apply() error {
+	if l.invalidWriter {
```

So (a) just call this "applied"/"hasApplied"/"invalid", and set it immediately after this early return. And (b) there's no need to set it via the SetInvalid method, we can just let this silently fail if garbage collection gets to the tmp table before the final apply